designer/restmodel/src/main/scala/pl/touk/nussknacker/restmodel/displayedgraph/DisplayableProcess.scala

```diff
@@ -49,22 +50,60 @@ import pl.touk.nussknacker.engine.graph.NodeDataCodec._
+case class ProcessProperties(additionalFields: ProcessAdditionalFields, propertiesType: String) {
 
-  def toMetaData(id: String): MetaData = MetaData(
-    id = id,
-    typeSpecificData = typeSpecificProperties,
-    additionalFields = additionalFields
-  )
-  val isSubprocess: Boolean = typeSpecificProperties.isSubprocess
+  def toMetaData(id: String): MetaData = MetaData(id = id, properties = additionalFields, propertiesType = propertiesType)
+
+  val isSubprocess: Boolean = additionalFields.extractTypeSpecificData(propertiesType)._1.isSubprocess
 
 }
 
-object ProcessProperties {
-  implicit val encodeProcessProperties: Encoder[ProcessProperties] =
-    Encoder.forProduct3("typeSpecificProperties", "isSubprocess", "additionalFields") { p =>
-    (p.typeSpecificProperties, p.isSubprocess, p.additionalFields)
+object ProcessProperties extends LazyLogging {
+
+  def apply(metaData: MetaData): ProcessProperties = {
+    ProcessProperties(metaData.typeSpecificData, metaData.additionalFields)
+  }
+
+  def apply(
+    typeSpecificProperties: TypeSpecificData,
+    additionalFields: Option[ProcessAdditionalFields] = None
+  ): ProcessProperties = {
+    val mergedProps = merge(additionalFields, typeSpecificProperties)
+    // We set the classname to make converting back from generic to typed easier
+    val typeSpecificClassName = typeSpecificProperties match {
+      case _: StreamMetaData          => "StreamMetaData"
+      case _: LiteStreamMetaData      => "LiteStreamMetaData"
+      case _: RequestResponseMetaData => "RequestResponseMetaData"
+      case _: FragmentSpecificData    => "FragmentSpecificData"
+      case _                          => throw new IllegalStateException("Type specific properties name not recognized.")
+    }
+    ProcessProperties(mergedProps, typeSpecificClassName)
   }
+
+  private def merge(properties: Option[ProcessAdditionalFields],
+                    typeSpecificData: TypeSpecificData): ProcessAdditionalFields = {
+    val fields = properties.getOrElse(ProcessAdditionalFields.empty)
+    // When merging we overwrite generic properties with TypeSpecificData. This would only happen if someone had
+    // a property with the same name was written as a generic property and in type specific properties. This shouldn't
+    // happen, but if it does, the data will heal on next save:
+    //  - if the value has a type that can be stored in TypeSpecificData it will be stored there and removed from
+    //    generic properties
+    //  - if the value has different type than in TypeSpecificData it will be stored in generic properties and in
+    //    TypeSpecificData it will be set as null
+    // This is a temporary solution until we get rid of TypeSpecificData completely.
+    typeSpecificData.toProperties
+      .filter(p => fields.properties.contains(p._1))
+      .foreach(p => logger.warn(s"Duplicate properties with the same name '${p._1}'. Overwriting additional property " +
```

Comment:
I assume that this will happen every time and we will have this warn during normal usage?  I would add this warn only when a value is different. WDYT?

Reply:
The 'toProperties' method returns a map of non-null properties. So if we have an invalid property (which in TypeSpcificData will be written as null), it won't be placed in the returned map. I will add this explanation to the comment - your assumption is definitely warranted.